11222 editable notice

danielfmiranda opened the pull request:
# Description

<!-- Describe the PR here -->

Link to sample test page: https://donate-foundation-s-11222-edit-walc5o.mofostaging.net/en/help/
Link to sample test page (CMS): https://foundation-s-11222-edit-walc5o.herokuapp.com/cms/pages/217/edit/
Related PRs/issues: #11222


This PR includes the following updates:
- A new streamfield block called `NoticeBlock`. This block contains 3 fields:
  - `image` 
  - `image_alt_text`
  - `text` - (a rich text field)
  - **[Note]** - While the image and image alt text fields are optional, this block includes custom validation in the clean() method to make sure that images have corresponding alt text.
- A new field on the `donatehelppage` called `notice`. This field allows users in the CMS to create a `NoticeBlock` for use in the template.
- Tests for the new `NoticeBlock`
- Migrations necessary for the new models/fields
- Updates to the `donatehelppage` template in order to render the new notice block


# Screenshots

With Image:
![Screenshot 2023-10-13 at 12-15-18 Donate Help](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/da095054-ba01-4d5a-a708-2f82a7752258)


Without Image:
![Screenshot 2023-10-13 at 12-15-50 Donate Help](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/3dd5cdda-56f9-4b2b-aa99-e547e4e00109)

No notice:
![Screenshot 2023-10-13 at 12-16-11 Donate Help](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/7f69ab73-7e5d-4a05-82e8-e806ea47c7ed)


CMS validation error (image but no alt text):
![Screenshot 2023-10-13 at 12-16-33 Editing Donate help page Donate Help - Wagtail](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/2e6e1751-1e33-43b7-a8f4-3c8bdccefce6)

CMS validation error (alt text but no image):
![Screenshot 2023-10-13 at 12-16-49 Editing Donate help page Donate Help - Wagtail](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/e023f634-b14f-4831-9c63-43cf9f2f7d7d)

jhonatan-lopes commented on network-api/networkapi/donate/migrations/0004_remove_donatehelppage_show_notice_flag.py:
Can we merge this migration with `0005`, please?

danielfmiranda replied:
Can do! As discussed over zoom, I have updated the migration files to be only one file, which creates the notice field and notice block.

jhonatan-lopes commented on network-api/networkapi/donate/tests/customblocks/test_notice_block.py:
```diff
@@ -0,0 +1,55 @@
+            "image_alt_text": "Alt text",
+            "text": rich_text.RichText("<p>Some content</p>"),
+        }
+        # If clean method does not raise ValidationError, the test will pass
```
I think it is better to be explicit here, rather than relying on something not happening...

danielfmiranda replied:
Thanks for the feedback Jhonatan! I have updated these tests with `assert` statements, and added a new one for the new `NoticeBlockFactory`

jhonatan-lopes commented on network-api/networkapi/donate/tests/customblocks/test_notice_block.py:
```diff
@@ -0,0 +1,55 @@
+from networkapi.donate.pagemodels.customblocks.notice_block import NoticeBlock
+
+
+class NoticeBlockTest(TestCase):
```
I think these tests would benefit from a `NoticeBlockFactory`. We could also add some tests for the page with the block...

danielfmiranda replied:
Updated using the 2 PR's you have sent me for reference! So I have updated the `DonateHelpPage` factory and tests to account for the new `NoticeBlockFactory`. And have also added a test for the factory as well.

jhonatan-lopes reviewed (commented):
Hey @danielfmiranda, it's all looking good, thanks! The only thing missing is a factory for the new block and to update the tests and the page factory to include the `NoticeBlockFactory`. You can find an example of that [on this PR](https://github.com/MozillaFoundation/foundation.mozilla.org/pull/11004/files)

danielfmiranda commented:
Hi @jhonatan-lopes thanks for the review and great feedback! I have implemented the `NoticeBlockFactory` as well as some updated tests. 

I did want to note that my new commits include about 5-6 more files than before. This is because with the addition of the new `NoticeBlockFactory`, I refactored the original factory file into multiple page-specific files, similar to the other apps in our codebase.

jhonatan-lopes commented on network-api/networkapi/donate/tests/customblocks/test_notice_block.py:
```diff
@@ -0,0 +1,98 @@
+        """
+        notice_block_factory_default_values = NoticeBlockFactory()
+
+        is_valid = self.notice_block.clean(notice_block_factory_default_values)
+
+        self.assertTrue(is_valid)
```
Much better! :+1:  thanks

jhonatan-lopes commented on network-api/networkapi/donate/tests/test_help_page.py:
```diff
@@ -64,3 +65,9 @@ def test_template(self):
+        """
+        Asserts that a 'notice' block was created in the 'notice' field by the factory.
+        """
+        self.assertEqual(self.donate_help_page.notice[0].block_type, "notice")
```
Nice

jhonatan-lopes reviewed (changes requested):
Hey Daniel, it's all looking good, but I think we just have a small issue with the factories being duplicated on `donate/factory.py`. Could you clean that up, please?

danielfmiranda commented on network-api/networkapi/donate/factory/customblocks/notice_block.py:
```diff
@@ -0,0 +1,20 @@
+description_faker = Faker("paragraphs", nb=2)
+
+
+class NoticeBlockFactory(wagtail_factories.StructBlockFactory):
+    class Meta:
+        model = NoticeBlock
+        exclude = ("description_text",)
+
+    image = SubFactory(wagtail_factories.ImageChooserBlockFactory)
+    image_alt_text = Faker("sentence", nb_words=4)
+    text = LazyAttribute(lambda o: RichText("".join([f"<p>{p}</p>" for p in o.description_text])))
+
+    # Lazy Values
+    description_text = description_faker
```
Sorry about that @jhonatan-lopes! Looks like I forgot to delete the original file 😅 I have deleted it and updated the PR for re-review. Thanks!

jhonatan-lopes reviewed (approved):
LGTM :+1: thanks @danielfmiranda
